src/main/scala/uk/pkerrigan/darwin/Message.scala

```diff
@@ -0,0 +1,63 @@
+package uk.pkerrigan.darwin
+
+import scala.xml.{Node, XML}
+
+sealed trait Time
+case class EstimatedTime(time: String) extends Time
+case class ActualTime(time: String) extends Time
+
+sealed trait Location {
+  def tiploc: String
+}
+case class Station(tiploc: String, arrivalTime: Time, departureTime: Time) extends Location
+case class Waypoint(tiploc: String, passTime: Time) extends Location
+
+sealed trait Message {
+  def origin: String
+}
+case class ScheduleUpdate(origin: String, runId: String, locations: List[Location]) extends Message
+
+object Message {
+  def parseMessage(message: String): Option[Message] =
+    (XML.loadString(message) \ "uR")
+      .headOption
+      .flatMap(this.parseUpdateResponse)
+
+  private def parseUpdateResponse(updateResponse: Node): Option[Message] =
+    updateResponse.child
+      .headOption
+      .flatMap(this.parseMessageType(updateResponse.attribute("updateOrigin").head.text)(_))
+
+  private def parseMessageType(origin: String)(messageContent: Node): Option[Message] =
+    messageContent.label match {
+      case "TS" => this.parseSchedule(origin)(messageContent)
+      case _ => None
+    }
+
+  private def parseSchedule(origin: String)(messageContent: Node): Option[ScheduleUpdate] = {
+    val locations = (messageContent \ "Location")
+      .map(this.parseLocation)
+      .toList
+    Some(ScheduleUpdate(origin, messageContent.attribute("rid").head.text, locations))
+  }
+
+  private def parseLocation(location: Node): Location = {
+    val time = this.getTime(location) _
+    (time("arr"), time("dep"), time("pass")) match {
+      case (Some(arr), Some(dep), None) => Station(location.attribute("tpl").head.text, arr, dep)
+      case (None, None, Some(pass)) => Waypoint(location.attribute("tpl").head.text, pass)
+      case _ => throw new Exception("Unknown location type")
```

Comment:
the unintended explosions may occur if you're doing toRight(throw new Exception) as opposed to toRight(new Exception) maybe? I've done that many a time.

And as for the List[Either] to Either[List] - it's a classic and you have a few choices

- use [cats ](http://typelevel.org/cats/)which provides a [.sequence function](http://eed3si9n.com/herding-cats/Traverse.html#sequence+function) if you're comfortable with typeclasses
- use foldLeft maybe?  psuedocode below 
- use [partition ](https://stackoverflow.com/questions/6489584/best-way-to-turn-a-lists-of-eithers-into-an-either-of-lists)as suggested by someone on stackoverflow 

so [foldLeft](https://commitlogs.com/2016/09/10/scala-fold-foldleft-and-foldright/)is basically how I tend to iterate through lists in Scala,

its like a reduce but lets you go through a list or similar and accumulate some kind of result based on the elements

```scala
locations.foldLeft[Either[Throwable, List[Location]](Right(List.empty)) { 
  case (err@Left(_), _) => err
  case (_, err@Left(_)) => err
  case (Right(prev), Right(cur)) => Right(prev :+ cur)
}
```